contracts/RockPaperScissors.sol

```diff
@@ -0,0 +1,320 @@
+    
+    enum Steps {
+        UNSTARTED, // 0
+        PLAYER_ONE_MOVED, // 1
+        PLAYER_TWO_MOVED, // 2
+        PLAYER_ONE_REVEALED // 3
+    }
+
+    enum GameResult {
+        DRAW, // 0
+        LEFTWIN, // 1
+        RIGHTWIN, // 2
+        INCORRECT // 3
+    }
+
+    enum Moves {
+        EMPTY, // 0
+        ROCK, // 1
+        PAPER, // 2
+        SCISSORS // 3
+    }
+
+    struct Game {
+        address playerOne;
+        Moves playerOneMove;
+        uint256 stake;
+        uint256 expiryDate;
+        Steps step;
+        address playerTwo;
+        bytes32 playerTwoMoveHash;
+    }
+
+    event PlayerOneMoves(
+        bytes32 indexed gameKey,
+        address indexed player,
+        uint256 stake,
+        uint256 amount
+    );
+
+    event PlayerTwoMoves(
+        bytes32 indexed gameKey,
+        address indexed player,
+        uint256 amount,
+        bytes32 moveHash,
+        uint256 expiryDate
+    );
+
+    event PlayerOneReveals(
+        bytes32 indexed gameKey,
+        address indexed player,
+        Moves move,
+        uint256 expiryDate
+    );
+
+    event PlayerTwoReveals(
+        bytes32 indexed gameKey,
+        address indexed player,
+        Moves move
+    );
+
+    event DrawRefund(
+        bytes32 indexed gameKey,
+        address indexed playerOne,
+        address indexed playerTwo,
+        uint256 playerOneRefund,
+        uint256 playerTwoRefund
+    );
+
+    event GamePayment(
+        bytes32 indexed gameKey,
+        address indexed player,
+        uint256 amount
+    );
+
+    event ForfeitPaid(
+        bytes32 indexed gameKey,
+        address indexed player,
+        uint256 amount
+    );
+
+    event WithDraw(
+        address indexed player,
+        uint amount
+    );
+
+    event PlayerOneEndsGame(
+        bytes32 indexed gameKey,
+        address indexed player,
+        uint256 amount
+    );
+
+    // The creator's move is embedded in the game key
+    function movePlayerOne(bytes32 gameKey, uint256 stake) external payable whenNotPaused {
+
+        require(gameKey != bytes32(0), "Game key hash is empty");
+        Game storage game = games[gameKey];
+        require(game.playerOne == NULL_ADDRESS, "Game key already used");
+
+        uint256 expiryDate = block.timestamp.add(FORFEIT_WINDOW);
+        game.expiryDate = expiryDate;
+        game.playerOne = msg.sender;
+        game.step = Steps.PLAYER_ONE_MOVED;
+        game.stake = stake;
+        depositStake(stake);
+
+        emit PlayerOneMoves(gameKey, msg.sender, stake, msg.value);
+    }
+
+    function movePlayerTwo(bytes32 gameKey, bytes32 moveHash) external payable whenNotPaused {
+
+        require(moveHash != bytes32(0), "Move hash is empty");
+        Game storage game = games[gameKey];
+        require(game.step == Steps.PLAYER_ONE_MOVED, INVALID_STEP_MSG);
+
+        uint256 expiryDate = block.timestamp.add(FORFEIT_WINDOW);
+        game.expiryDate = expiryDate;
+        game.playerTwo = msg.sender;
```

Comment:
Now that we are reaching the end, it is worth reminding you that you had player 2 hash their move too because you wanted to keep open a created game.
If you had player2 defined at createGame, then player 2 could send the move in the clear.
Also keep in mind that with an open game like that, if there are more "player2s" than "player1s", then a lot of player2s, all those who try to join the same open game, will see their transaction revert. Not a good look.

Reply:
I recall my original design from around 22/03 was to create the game with both players identified in an initialise() function created by the contract owner. The players would be matched off-chain before it is initiated.

I am not sure why I subsequently changed it to allow players to join an open game. I think I assumed that as soon as a second player would join then the event would emit and the UI would block other players. However, I appreciate there would be a delay between a second player joining and the the event being emitted which means that other players will try to join a game which appears to be open but isn't due to the event emitting lag. I understand that this could  be mitigated against by only playing the game via another layer e.g. a web3 enabled app which has its own storage that records player activity and controls who can join a game.

Reply:
Yes. If you wanted to "stay longer" on contracts, you could imagine a match-making contract that would "find" an available game for about the same stake. Only if there was no such game left, would the tx revert. By "find", it is not really a search algorithm because we want to stay O(1).
That could probably be implemented with a sort of queue.